src/libos/crates/host-socket/src/stream/mod.rs

```diff
@@ -185,6 +185,47 @@ impl<A: Addr, R: Runtime> StreamSocket<A, R> {
             .unregister_observer(observer)
             .ok_or_else(|| errno!(ENOENT, "the observer is not registered"))
     }
+
+    pub fn addr(&self) -> Result<A> {
+        let state = self.state.read().unwrap();
+        match &*state {
+            State::Init(init_stream) => init_stream.common().addr().map_or_else(
+                || init_stream.common().get_addr_from_host(),
+                |addr| Ok(addr),
+            ),
+            State::Connect(connecting_stream) => connecting_stream.common().addr().map_or_else(
+                || connecting_stream.common().get_addr_from_host(),
+                |addr| Ok(addr),
+            ),
+            State::Connected(connected_stream) => connected_stream.common().addr().map_or_else(
+                || {
+                    let addr = connected_stream.common().get_addr_from_host()?;
+                    connected_stream.common().set_addr(&addr);
+                    Ok(addr)
+                },
+                |addr| Ok(addr),
+            ),
+            State::Listen(listener_stream) => listener_stream.common().addr().map_or_else(
+                || {
+                    let addr = listener_stream.common().get_addr_from_host()?;
+                    listener_stream.common().set_addr(&addr);
+                    Ok(addr)
+                },
+                |addr| Ok(addr),
+            ),
```

Comment:
The cases of connected and listening sockets can be combined to avoid code duplication and make the fact that their logic is the same explicit.